plugins/sqs_worker/worker.go

```diff
@@ -0,0 +1,154 @@
+//go:build sqs && sqs_worker
+// +build sqs,sqs_worker
+
+package sqs_worker
+
+import (
+	"context"
+	"github.com/aws/aws-sdk-go/aws"
+	aws_sqs "github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/shoplineapp/go-app/plugins"
+	"github.com/shoplineapp/go-app/plugins/logger"
+	"github.com/shoplineapp/go-app/plugins/sqs"
+	"go.uber.org/fx"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type AwsSqsWorker struct {
+	TopicMgr    *sqs.AwsTopicManager
+	Logger      *logger.Logger
+	HandlerMaps map[string]EventHandlerInterface
+}
+
+type awsMessage struct {
+	topicName string
+	*aws_sqs.Message
+}
+
+var enabled = true
+var started = false
+var sqsMaxMessages = int64(1)
+
+func init() {
+	plugins.Registry = append(plugins.Registry, NewAwsSqsWorker)
+}
+
+func (w *AwsSqsWorker) SetRegion(region string) {
+	w.TopicMgr.SetRegion(region)
+}
+
+func (w *AwsSqsWorker) Serve() {
+	chnMessages := make(chan *awsMessage, len(w.HandlerMaps))
+
+	for _, handlerInterface := range w.HandlerMaps {
+		topic := w.TopicMgr.GetTopic(handlerInterface.Topic().Name)
+		go w.messageLoop(chnMessages, topic)
+		w.Logger.WithFields(log.Fields{"topic": topic.Name, "queueArn": topic.Arn}).Info("Listening on SQS queue")
+	}
+
+	go func() {
+		for message := range chnMessages {
+			w.handleMessage(message)
+		}
+	}()
+
+	started = true
+}
+
+func (w *AwsSqsWorker) messageLoop(chn chan<- *awsMessage, topic *sqs.Topic) {
+	for enabled {
+		output, err := topic.ReceiveMessage(&aws_sqs.ReceiveMessageInput{
+			AttributeNames: []*string{
+				aws.String(aws_sqs.MessageSystemAttributeNameSentTimestamp),
+			},
+			MessageAttributeNames: []*string{
+				aws.String(aws_sqs.QueueAttributeNameAll),
+			},
+			QueueUrl:            &topic.Arn,
+			MaxNumberOfMessages: aws.Int64(sqsMaxMessages),
+			VisibilityTimeout:   aws.Int64(20), // 20 seconds
+			WaitTimeSeconds:     aws.Int64(0),
+		})
+
+		if err != nil {
+			w.Logger.WithFields(log.Fields{"error": err}).Error("Failed to fetch sqs message")
+		}
+
+		for _, message := range output.Messages {
+			chn <- &awsMessage{
+				topicName: topic.Name,
+				Message:   message,
+			}
+		}
+	}
+}
+
+func (w *AwsSqsWorker) handleMessage(message *awsMessage) {
+	defer func() {
+		if r := recover(); r != nil {
+			err := r.(error)
+			w.Logger.WithFields(log.Fields{"error": err}).Error("Failed to invoke event")
+
+			topic := w.TopicMgr.GetTopic(message.topicName)
+			deleteMessageInput := aws_sqs.DeleteMessageInput{QueueUrl: &topic.Arn, ReceiptHandle: message.ReceiptHandle}
+			_, _ = topic.DeleteMessage(&deleteMessageInput)
```

Comment:
why do we delete the message when a panic is occurred, i think the default behaviour should be ignoring the error and let the sqs re-send the message for retry

Reply:
and i think we should call OnEvent there if possible, this should be part of handler error